smarts/core/controllers/imitation_controller.py

```diff
@@ -0,0 +1,78 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+import math
+import numpy as np
+from dataclasses import replace
+
+from smarts.core.chassis import AckermannChassis, BoxChassis
+from smarts.core.utils.math import fast_quaternion_from_angle, radians_to_vec
+
+
+class ImitationController:
+    @classmethod
+    def perform_action(cls, dt, vehicle, action):
+        if isinstance(action, (int, float)):
+            # special case:  setting the initial speed
+            if isinstance(vehicle.chassis, BoxChassis):
+                vehicle.control(vehicle.pose, action)
+            elif isinstance(vehicle.chassis, AckermannChassis):
+                vehicle.chassis.speed = action  # hack that calls control internally
+            return
+
+        assert isinstance(action, (list, tuple)) and len(action) == 2
+        # acceleration in m/s^2, angluar_velocity in rad/s
+        acceleration, angular_velocity = action
+
+        if isinstance(vehicle.chassis, BoxChassis):
+            target_speed = vehicle.speed + acceleration * dt
+            target_heading = (vehicle.heading + angular_velocity * dt) % (2 * math.pi)
+            speed_x, speed_y = radians_to_vec(target_heading)
+            new_position = [
+                vehicle.position[0] + dt * speed_x * target_speed,
+                vehicle.position[1] + dt * speed_y * target_speed,
+                vehicle.position[2],
+            ]
+            new_pose = replace(
+                vehicle.pose,
+                orientation=fast_quaternion_from_angle(target_heading),
+                position=new_position,
+            )
+            vehicle.control(new_pose, target_speed)
+
+        # TAI: consider making this a dynamic_action_space, like...
+        elif isinstance(vehicle.chassis, AckermannChassis):
+            mass = vehicle.chassis.mass_and_inertia[0]  # in kg
+            if acceleration >= 0:
+                # necessary torque is N*m = kg*m*acceleration
+                torque_ratio = mass / vehicle.chassis.max_torque
+                throttle = np.clip(acceleration * torque_ratio, 0, 1)
```

Comment:
Just a minor thing, since the throttle is dimensionless, may be we should use: 
`torque_ratio = mass / (4*wheel_radius*vehicle.chassis.max_torque)`

Reply:
I think that in the chassis throttle value (between `0` and `1`) is scaled up to be between `0` nd `max_torque`.  That's why I did it this way.  It seems that is we also normalize here by `4 * wheel_radius`, we're limiting the maximum amount of torque we can provide to the vehicle (and so the maximum acceleration) by something like 80% of its available range.  I didn't think we'd want to do that. 
 
Or am I misunderstanding things?

Reply:
When we use `AckermannChassis` and we go through the bullet. In `control` method of `AckermannChassis`, a list of throttles for each wheel will be created(if we pass a scalar throttle), thus if we want to produce specific acceleration with bullet the equation of motion dictates:

`4*wheel_radius*throttle*max_torque=mass*acceleration`

from which we obtain:

`throttle=mass*acceleration/(4*wheel_radius*max_torque)`

Reply:
Ah ha, I see that now.  Thank you!  I will make the change.